test/e2e/network/topology.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/network/common"
+)
+
+var _ = common.SIGDescribe("[Feature:TopologyAwareHints] Topology Aware Hints", func() {
+
+	f := framework.NewDefaultFramework("topology-hints")
+
+	var cs clientset.Interface
+
+	ginkgo.BeforeEach(func() {
+		cs = f.ClientSet
+	})
+
+	ginkgo.It("Services with topology annotation should forward traffic to nodes on the same zone", func() {
+		namespace := f.Namespace.Name
+		serviceName := "svc-topology"
+		port := 80
+
+		jig := e2eservice.NewTestJig(cs, namespace, serviceName)
+		nodes, err := e2enode.GetBoundedReadySchedulableNodes(cs, e2eservice.MaxNodesForEndpointsTests)
+		framework.ExpectNoError(err)
+		// select one node in a specific to create a pod to generate the traffic
+		nodeName := nodes.Items[0].Name
+		nodeZone, ok := nodes.Items[0].Labels["topology.kubernetes.io/zone"]
+		if !ok {
+			e2eskipper.Skipf("Skipping because nodes don't have the topology zone information")
+		}
+
+		ginkgo.By("creating an annotated service with no endpoints and topology aware annotation")
+		svc, err := jig.CreateTCPServiceWithPort(func(svc *v1.Service) {
+			svc.Annotations = map[string]string{"service.kubernetes.io/topology-aware-hints": "auto"}
+			svc.Spec.Ports = []v1.ServicePort{
+				{Port: int32(port), Name: "http", Protocol: v1.ProtocolTCP, TargetPort: intstr.FromInt(9376)},
+			}
+		}, int32(port))
+		framework.ExpectNoError(err)
+		svcIP := svc.Spec.ClusterIP
+
+		ginkgo.By("creating backend pods for the service on each node" + serviceName)
+		// hints are only generated if there is a reasonable distribution of endpoints
+		// TODO(aojea): revisit, this number was obtained with KIND in a cluster with 3 nodes
```

Comment:
How can we ensure that this test only runs in multi-zone environments?

Reply:
let's keep this for the kube-proxy fix